include/bio/plain_io/misc.hpp

```diff
@@ -0,0 +1,122 @@
+
+/*!\file
+ * \brief Provides utility types for bio::plain_io::reader and bio::plain_io::writer.
+ * \author Hannes Hauswedell <hannes.hauswedell AT decode.is>
+ */
+
+#pragma once
+
+#include <stdexcept>
+#include <string_view>
+#include <vector>
+
+#include <bio/platform.hpp>
+
+namespace bio::plain_io
+{
+
+/*!\addtogroup plain_io
+ * \{
+ */
+
+//!\brief The value type of bio::plaintext_file_input if every line is split into fields.
+struct record
+{
+    //!\brief The entire line (exluding EOL characters but including delimiters).
+    std::string_view              line;
+    //!\brief A range of the individual fields (without delimiters or EOL characters).
+    std::vector<std::string_view> fields;
+};
+
+//!\brief Option to switch between reading-by-line and splitting a line into fields.
+enum class record_kind
+{
+    line,           //!< Only the line is provided.
+    line_and_fields //!< The line is provided and also individual fields (bio::plaintext_record).
+};
+
+/*!\brief A helper for specifying the header of a bio::plaintext_file_input.
+ * \tparam record_kind Whether to split lines on delimiter (e.g. TSV files) or not.
+ *
+ * \details
+ *
+ * TODO explain how this can be used easily
+ */
+class header_kind
+{
+private:
+    //!\brief The state of the variable, encoded as the value of the char or the special values -200 and -300.
+    int state = -200;
+
+public:
+    //!\brief The state representing "no header".
+    static constexpr struct
+    {
+    } none{};
+    //!\brief The state representing "first line is header".
+    static constexpr struct
+    {
+    } first_line{};
+
+    //!\brief Type of the state representing "all lines that start with character X".
+    struct starts_with
+    {
+        //!\privatesection
+        //!\brief Store the chracter.
+        char c;
+    };
+
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    constexpr header_kind() noexcept                    = default;             //!< Defaulted.
+    constexpr header_kind(header_kind const &) noexcept = default;             //!< Defaulted.
+    constexpr header_kind(header_kind &&) noexcept      = default;             //!< Defaulted.
+    constexpr header_kind & operator=(header_kind const &) noexcept = default; //!< Defaulted.
+    constexpr header_kind & operator=(header_kind &&) noexcept = default;      //!< Defaulted.
+
+    //!\brief Initialise to the "none"-state.
+    constexpr header_kind(decltype(none)) noexcept : state{-200} {}
+    //!\brief Initialise to the "first_line"-state.
+    constexpr header_kind(decltype(first_line)) noexcept : state{-300} {}
+    //!\brief Initialise to the "starts_with"-state and a given character.
+    constexpr header_kind(starts_with s) noexcept : state{s.c} {}
+    //!\}
+
+    /*!\name Functions for retrieving the state.
+     * \{
+     */
+    //!\brief Whether this object is in the "none"-state.
+    constexpr bool is_none() const noexcept { return state == -200; }
```

Comment:
I would add a variable `constexpr int none = -200;` to the class, such that this and other places become more expressive.
```suggestion
    constexpr bool is_none() const noexcept { return state == none; }
```

Reply:
There already is a member variable called "none".

Reply:
choose a different name?

Reply:
I did!